db/write_batch_internal.h

```diff
@@ -260,4 +260,173 @@ class LocalSavePoint {
 #endif
 };
 
+template <typename Checker>
+class TimestampAssignerBase : public WriteBatch::Handler {
+ public:
+  explicit TimestampAssignerBase(WriteBatch::ProtectionInfo* prot_info,
+                                 Checker&& checker)
+      : prot_info_(prot_info), checker_(std::move(checker)) {}
+
+  ~TimestampAssignerBase() override {}
+
+  Status PutCF(uint32_t cf, const Slice& key, const Slice&) override {
+    Status s = AssignTimestamp(cf, key);
+    ++idx_;
+    return s;
+  }
+
+  Status DeleteCF(uint32_t cf, const Slice& key) override {
+    Status s = AssignTimestamp(cf, key);
+    ++idx_;
+    return s;
+  }
+
+  Status SingleDeleteCF(uint32_t cf, const Slice& key) override {
+    Status s = AssignTimestamp(cf, key);
+    ++idx_;
+    return s;
+  }
+
+  Status DeleteRangeCF(uint32_t cf, const Slice& begin_key,
+                       const Slice&) override {
+    Status s = AssignTimestamp(cf, begin_key);
+    ++idx_;
+    return s;
+  }
+
+  Status MergeCF(uint32_t cf, const Slice& key, const Slice&) override {
+    Status s = AssignTimestamp(cf, key);
+    ++idx_;
+    return s;
+  }
+
+  Status PutBlobIndexCF(uint32_t cf, const Slice& key, const Slice&) override {
+    Status s = AssignTimestamp(cf, key);
+    ++idx_;
+    return s;
+  }
+
+  Status MarkBeginPrepare(bool) override { return Status::OK(); }
+
+  Status MarkEndPrepare(const Slice&) override { return Status::OK(); }
+
+  Status MarkCommit(const Slice&) override { return Status::OK(); }
+
+  Status MarkRollback(const Slice&) override { return Status::OK(); }
+
+ protected:
+  virtual Status AssignTimestamp(uint32_t cf, const Slice& key) = 0;
```

Comment:
We could consider using CRTP to avoid the need for the additional virtual method calls.

Reply:
Done